wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/jpa/JpaUserDevfileDao.java

```diff
@@ -0,0 +1,420 @@
+import com.google.common.base.Supplier;
+import com.google.common.collect.ImmutableList;
+import com.google.inject.persist.Transactional;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.StringJoiner;
+import java.util.stream.Collectors;
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+import javax.persistence.EntityManager;
+import javax.persistence.TypedQuery;
+import org.eclipse.che.account.event.BeforeAccountRemovedEvent;
+import org.eclipse.che.account.shared.model.Account;
+import org.eclipse.che.account.spi.AccountDao;
+import org.eclipse.che.api.core.ConflictException;
+import org.eclipse.che.api.core.NotFoundException;
+import org.eclipse.che.api.core.Page;
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.core.model.workspace.devfile.UserDevfile;
+import org.eclipse.che.api.core.notification.EventService;
+import org.eclipse.che.api.devfile.server.UserDevfileManager;
+import org.eclipse.che.api.devfile.server.event.BeforeDevfileRemovedEvent;
+import org.eclipse.che.api.devfile.server.model.impl.UserDevfileImpl;
+import org.eclipse.che.api.devfile.server.spi.UserDevfileDao;
+import org.eclipse.che.commons.lang.Pair;
+import org.eclipse.che.core.db.cascade.CascadeEventSubscriber;
+import org.eclipse.che.core.db.jpa.DuplicateKeyException;
+import org.eclipse.che.core.db.jpa.IntegrityConstraintViolationException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** JPA based implementation of {@link UserDevfileDao}. */
+@Singleton
+@Beta
+public class JpaUserDevfileDao implements UserDevfileDao {
+  private static final Logger LOG = LoggerFactory.getLogger(JpaUserDevfileDao.class);
+
+  protected final Provider<EntityManager> managerProvider;
+  protected final AccountDao accountDao;
+  protected final EventService eventService;
+
+  public static final List<Pair<String, String>> DEFAULT_ORDER =
+      ImmutableList.of(new Pair<>("id", "ASC"));
+
+  @Inject
+  public JpaUserDevfileDao(
+      Provider<EntityManager> managerProvider, AccountDao accountDao, EventService eventService) {
+    this.managerProvider = managerProvider;
+    this.accountDao = accountDao;
+    this.eventService = eventService;
+  }
+
+  @Override
+  public UserDevfile create(UserDevfile userDevfile)
+      throws ConflictException, ServerException, NotFoundException {
+    requireNonNull(userDevfile);
+    try {
+      Account account = accountDao.getByName(userDevfile.getNamespace());
+      UserDevfileImpl userDevfileImpl = new UserDevfileImpl(userDevfile, account);
+      doCreate(userDevfileImpl);
+      return userDevfileImpl;
+    } catch (DuplicateKeyException ex) {
+      throw new ConflictException(
+          format(
+              "Devfile with name '%s' already exists in current account '%s'",
```

Comment:
In the case of organization, it's not really current by specified.
```suggestion
              "Devfile with name '%s' already exists in the specified account '%s'",
```